python/src/iceberg/transforms.py

```diff
@@ -0,0 +1,520 @@
+        "day": lambda t1, t2: (t1 - t2).days,
+        "hour": lambda t1, t2: int((t1 - t2).total_seconds() / 3600),
+    }
+
+    def __init__(self, transform_type: Type, name: str):
+        if name not in Time._VALID_TIME_GRANULARITY.get(transform_type, {}):
+            raise ValueError(f"Cannot transform type: {transform_type} by {name}")
+
+        super().__init__(
+            name, f"transforms.{name}(transform_type={repr(transform_type)})"
+        )
+        self._type = transform_type
+        self._name = name
+
+    def apply(self, value: int) -> int:
+        if self._name == "day" and self._type == DateType:
+            return value
+
+        if self._type == DateType:
+            value_time = datetime.utcfromtimestamp(value * 86400)
+        else:
+            value_time = datetime.utcfromtimestamp(value / 1000000)
+
+        return Time._DIFF_MAP[self._name](value_time, Transform._EPOCH)
+
+    def can_transform(self, target: Type) -> bool:
+        if self._type == DateType:
+            return target == DateType
+        else:  # self._type is either TimestampType or TimestamptzType
+            return target == TimestampType or target == TimestamptzType
+
+    def get_result_type(self, source_type: Type) -> Type:
+        if self._name == "day":
+            return DateType
+
+        return IntegerType
+
+    def preserve_order(self) -> bool:
+        return True
+
+    def satisfy_order(self, other: Transform) -> bool:
+        if self == other:
+            return True
+
+        if isinstance(other, Time):
+            return getattr(Time._TIME_ORDER, self._name) <= getattr(
+                Time._TIME_ORDER, other._name
+            )
+
+        return False
+
+    def to_human_string(self, value: int) -> str:
+        if value is None:
+            return "null"
+
+        if self._name == "year":
+            return "{0:0=4d}".format(Transform._EPOCH.year + value)
+        elif self._name == "month":
+            return "{0:0=4d}-{1:0=2d}".format(
+                Transform._EPOCH.year + int(value / 12), 1 + int(value % 12)
+            )
+        elif self._name == "day":
+            return Transform._human_day(value)
+        else:  # self._name == "hour":
+            dt = Transform._EPOCH + timedelta(hours=value)
+            return "{0:0=4d}-{1:0=2d}-{2:0=2d}-{3:0=2d}".format(
+                dt.year, dt.month, dt.day, dt.hour
+            )
+
+    def dedup_name(self) -> str:
+        return "time"
+
+
+class Identity(Transform):
+    _HUMAN_STRING_MAP = {
+        DateType: lambda v: Transform._human_day(v),
+        TimeType: lambda v: f"{(Transform._EPOCH + timedelta(microseconds=v)).time()}",
+        TimestampType: lambda v: (
+            Transform._EPOCH + timedelta(microseconds=v)
+        ).isoformat(),
+        TimestamptzType: lambda v: pytz.timezone("UTC")
+        .localize(Transform._EPOCH + timedelta(microseconds=v))
+        .strftime("%Y-%m-%dT%H:%M:%S.%fZ"),
```

Comment:
Formatting of lambdas also doesn't work here. I think it would be better to refer to methods.

Reply:
SG, moved them to transform_util module methods.